src/lua_test.c

```diff
@@ -0,0 +1,112 @@
+/**
+ * Copyright (C) 2024 Kentaro Hayashi
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ * Authors:
+ *   Kentaro Hayashi <kenhys at xdump.org>
+ **/
+
+#include "collectd.h"
+#include "lua.c"
+#include "testing.h"
+#include "utils_lua.c"
+
+static lua_State *L = NULL;
+static char config_path[PATH_MAX] = {0};
+const char *module_key_value_config = "<Plugin lua>\n"
+                                      "  Script \"example.lua\"\n"
+                                      "  <Module>\n"
+                                      "    Key \"Value\"\n"
+                                      "  </Module>\n"
+                                      "</Plugin>\n";
+
+void test_setup_config(const char *config) {
+  int fd = 0;
+  FILE *fp = NULL;
+  {
+    /* create temporary .conf */
+    strncpy(config_path, "lua-XXXXXX.conf", strlen("lua-XXXXXX.conf") + 1);
+    fd = mkstemps(config_path, /* .conf */ 5);
+    fp = fdopen(fd, "w+");
+    fprintf(fp, "%s", config);
+    fclose(fp);
+  }
+  {
+    /* create empty Lua script which is loaded via lua_config */
+    fp = fopen("example.lua", "w+");
+    fclose(fp);
+  }
+}
+
+void test_setup(void) {
+  L = luaL_newstate();
+  luaL_openlibs(L);
+}
+
+void test_teardown(void) { lua_close(L); }
+
+DEF_TEST(luaC_pushoconfigitem) {
+  test_setup_config(module_key_value_config);
+
+  oconfig_item_t *root = oconfig_parse_file(config_path);
+  oconfig_item_t *plugin = root->children;
+
+  /*
+    lua_config internally calls luaC_pushoconfigitem
+    and lua_State context is stored in each script.
+    <Module> will be mapped to Lua table:
+    { Key => Value } so check the value of "Key".
+  */
+  lua_config(plugin);
+  L = scripts->lua_state;
+  lua_getglobal(L, "example.lua");
+  lua_getfield(L, -1, "Key");
+  EXPECT_EQ_STR("Value", luaL_checkstring(L, -1));
+
+  oconfig_free(root);
+  unlink(config_path);
+  unlink("example.lua");
```

Comment:
Please move this to `test_teardown()`.

Reply:
These are outcome of test_setup_config, so I doubt it should be.

Reply:
That's my point: the environment that is set up in the setup() function, should be cleaned up in the teardown() function. The test code itself shouldn't have to worry about whether files were created and where.

It looks like there are multiple separate concerns here:

* Creating the config file -> `test_setup_config()`
* Setting up the Lua interpreter -> `test_setup()`

I suggest that:

1.  We rename `test_setup()` and `test_teardown()` to `setup_lua()` and `teardown_lua()` respectively.
2.  We pair `setup_config()` with a `teardown_config()` function.